skcosmo/sample_selection/_voronoi_fps.py

```diff
@@ -0,0 +1,243 @@
+
+        super()._init_greedy_search(X, y, n_to_select)
+
+    def _continue_greedy_search(self, X, y, n_to_select):
+        """Continues the search. Prepares an array to store the selected
+        features."""
+
+        super()._continue_greedy_search(X, y, n_to_select)
+
+        n_pad = n_to_select - self.n_selected_
+
+        self.dSL_ = np.pad(self.dSL_, (0, n_pad), "constant", constant_values=0)
+
+    def _get_active(self, X, last_selected):
+        """
+        Finds the indices of the Voronoi cells that might change due to the fact
+        that a new point has been selected (index is last_selected,
+        because we haven't called super() yet, self.n_selected_ has not been
+        incremented yet)
+
+        This function calculates the distances between the last selected point
+        and the previously selected points. Next, the list of points to which
+        the distance is to be calculated is shortened.
+
+        These are the points for which need to recompute distances.
+        Let:
+        L is the last point selected;
+        S are the selected points from before this iteration;
+        X are the candidates;
+        The logic here is that we want to check if d(XL) can be smaller than
+        min(d(X,S)) (which is stored in self.haussdorf_)
+        now, if a point belongs to the Voronoi cell of S then
+        min(d(X,S_i))=d(X,S). Triangle inequality implies that
+        d(S,L) < |d(X,S) + d(L,X)| so we just need to check if
+        |d(X,S) - d(S,L)|>= d(X,S) to know that we don't need to check X.
+        but |d(X,S) - d(S,L)|^2>= d(X,S)^2 if and only if d(S,L)/2 > d(S,X)
+        """
+
+        if not hasattr(self, "n_selected_") or self.n_selected_ == 0:
+            return np.arange(X.shape[0], dtype=int)
+
+        else:
+            self.dSL_[: self.n_selected_] = (
+                self.norms_[self.selected_idx_[: self.n_selected_]]
+                + self.norms_[last_selected]
+                - 2 * (self.X_selected_[: self.n_selected_] @ X[last_selected].T)
+            ) * 0.25
+            # calculation in a single block
+
+            active_points = np.where(
+                self.dSL_[self.vlocation_of_idx] < self.haussdorf_
+            )[0]
+
+            return active_points
+
+    def _update_post_selection(self, X, y, last_selected):
+        """
+        Saves the most recently selected feature and increments the feature counter
+        """
+
+        if self._axis == 1:
+            self.X_selected_[:, self.n_selected_] = np.take(
+                X, last_selected, axis=self._axis
+            )
+        else:
+            self.X_selected_[self.n_selected_] = np.take(
+                X, last_selected, axis=self._axis
+            )
+
+            if hasattr(self, "y_selected_"):
+                self.y_selected_[self.n_selected_] = y[last_selected]
+
+        self.selected_idx_[self.n_selected_] = last_selected
+        self.n_selected_ += 1
+
+    def score(self, X, y=None):
```

Comment:
Why do we this, rather than redefine the `_update_haussdorf` from the superclass?

Reply:
But the superclass has `_update_haussdorf`... I see this as the much more circuitous route, as it requires the repetition of the `_update_post_selection` class all the way from `GreedySelector`

Reply:
Yes, but then again, otherwise we would have to make up _update_haussdorf out of nothing and, as a consequence, make a bunch of edits to other functions. I honestly just don't understand what for.

Reply:
If you don't define `_update_haussdorf`, it will inherit from FPS regardless, so it is *not* out of nothing. I'm really against the re-defining of `_update_post_selection`, which was overwritten by FPS from GreedySelector. From an inheritance standpoint, we should not be re-writing a method of our direct superclass to that of the super-superclass. Programatically, we should be putting this within a `_update_haussdorf`. Sorry, but I'm going to be firm on this.

Reply:
I will insist on it, too. The problem is that `_update_haussdorf` superclass is useless in `Voronoi_FPS`. I had to take `_update_post_selection` from` GreedySelector` because `_update_haussdorf` in `Voronoi_FPS` is just not needed. We  can of course make a cheat and write something like `def _update_haussdorf(self, X, y, last_selected): pass`, but this is already some kind of crutch.

Reply:
I know why you chose to re-write from GreedySelector, but it is _very_ bad coding practice to do it this way. Re-write the `_update_haussdorf` to do what `score` does and eliminate `_update_post_selection`.

Reply:
In my opinion, it is also bad practice to create unnecessary function calls, which not only make no sense, but also slow down the algorithm. In my opinion, the solution I proposed is simpler and clearer. Besides, there is another solution - to rewrite `FPS` and remove `_update_haussdorf` from it. This function didn't exist before, and it is much easier to make corrections in the `FPS` (because the algorithm is simpler).

Reply:
Through negotiation, a great solution was reached - just remove inheritance from `FPS` in `Voronoi_FPS`.